src/accounts_refactor.rs

```diff
@@ -0,0 +1,339 @@
+// Copyright (c) 2022 Espresso Systems (espressosys.com)
+// This file is part of the Seahorse library.
+
+// This program is free software: you can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software Foundation, either version 3 of the License, or (at your option) any later version.
+// This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details.
+// You should have received a copy of the GNU General Public License along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+//! The accounts module.
+//!
+//! This module defines [Account], [AccountEditor], and [Accounts], which provide CURD (create, read,
+//! update, and delete) operations, with the use of [KeyValueStore] to control the accounts resource.
+
+use crate::{
+    events::{EventSource, LedgerEvent},
+    key_scan::BackgroundKeyScan,
+    key_value_store::KeyValueStore,
+    txn_builder::RecordInfo,
+    KeystoreError,
+};
+use chrono::{DateTime, Local};
+use jf_cap::{
+    keys::{FreezerKeyPair, FreezerPubKey, UserAddress, UserKeyPair, ViewerKeyPair, ViewerPubKey},
+    structs::AssetCode,
+    MerkleCommitment,
+};
+use primitive_types::U256;
+use reef::Ledger;
+use serde::{de::DeserializeOwned, Deserialize, Serialize};
+use std::collections::HashMap;
+use std::hash::Hash;
+use std::ops::{Deref, DerefMut};
+
+/// Keys with a public part.
+pub trait KeyPair: Clone + Send + Sync {
+    type PubKey: Clone + DeserializeOwned + Hash + Eq + Serialize;
+    fn pub_key(&self) -> Self::PubKey;
+}
+
+impl KeyPair for ViewerKeyPair {
+    type PubKey = ViewerPubKey;
+
+    fn pub_key(&self) -> Self::PubKey {
+        self.pub_key()
+    }
+}
+
+impl KeyPair for FreezerKeyPair {
+    type PubKey = FreezerPubKey;
+
+    fn pub_key(&self) -> Self::PubKey {
+        self.pub_key()
+    }
+}
+
+impl KeyPair for UserKeyPair {
+    // The PubKey here is supposed to be a conceptual "primary key" for looking up UserKeyPairs. We
+    // typically want to look up UserKeyPairs by Address, not PubKey, because if we have a PubKey we
+    // can always get an Address to do the lookup.
+    type PubKey = UserAddress;
+
+    fn pub_key(&self) -> Self::PubKey {
+        self.address()
+    }
+}
+
+#[derive(Clone, Deserialize, Serialize)]
+#[serde(bound = "Key: DeserializeOwned + Serialize")]
+pub struct Account<L: Ledger, Key: KeyPair> {
+    /// The account key.
+    key: Key,
+    /// The account description.
+    pub description: String,
+    /// Whether the account is used.
+    pub used: bool,
+    /// Optional ledger scan.
+    scan: Option<BackgroundKeyScan<L>>,
+    /// The list of asset codes of the account.
+    assets: Vec<AssetCode>,
+    /// The list of records of the account.
+    records: Vec<RecordInfo>,
```

Comment:
I can update this when I put up the PR for the Records Model

Reply:
Isn't this going to end up duplicating the details of each record (stored once in `Records` and once in `Accounts`)? I think this should be implemented as a join at the controller layer (which for now means a standalone function in `txn_builder.rs` or `lib.rs`), conceptually something like `AccountsController::records(&self) -> impl  Iterator<Item = RecordInfo>`, which does something like `model.records.iter().filter(|r| r.pub_key() == self.pub_key())`.

Reply:
We can also have `Vec<RecordUid>` or something (only storing the primary key, so we don't duplicate the storage of the actual record info). But I think it is redundant since we can always just filter records for this account's public key

Reply:
Yeah, I was thinking we could have `Vec<u64>` where `u64` is the transaction UID, so it's consistent with the format of `assets` that they both use the primary keys of the corresponding resources. When we have the controller layer, we can just remove the `assets` and `records` fields from `Account`.

Reply:
After more thinking, I think `assets` and `records` should just be removed from `Account`. The interactions among these resources should really be handled by the controller.

Reply:
That seems fine. We'll just need to make sure that whenever we add or a remove a record in `Records`, we also update `records` and `assets` here

Reply:
Oops your comment snuck in in front of mine. My comment applies to your previous one, obviously we don't have to worry about updating these fields if we delete them